persistence/src/main/scala/hmda/persistence/processing/SubmissionFSM.scala

```diff
@@ -0,0 +1,261 @@
+    case SubmissionParsing(s) => currentData.update(s)
+    case SubmissionParsed(s) => currentData.update(s)
+    case SubmissionParsedWithErrors(s) => currentData.update(s)
+    case SubmissionValidating(s) => currentData.update(s)
+    case SubmissionValidated(s) => currentData.update(s)
+    case SubmissionValidatedWithErrors(s) => currentData.update(s)
+  }
+
+  startWith(Idle, EmptySubmissionData)
+
+  when(Idle) {
+    case Event(Create, _) =>
+      goto(Created) applying SubmissionCreated(Submission(submissionId, hmda.model.fi.Created))
+
+  }
+
+  when(Created) {
+    case Event(StartUpload, _) =>
+      val status = hmda.model.fi.Uploading
+      updateStatus(status)
+      goto(Uploading) applying SubmissionUploading(Submission(submissionId, status))
+
+  }
+
+  when(Uploading) {
+    case Event(CompleteUpload, _) =>
+      val status = hmda.model.fi.Uploaded
+      updateStatus(status)
+      goto(Uploaded) applying SubmissionUploaded(Submission(submissionId, status))
+    case Event(_, data) =>
+      stay replying data
+  }
+
+  when(Uploaded) {
+    case Event(StartParsing, _) =>
+      val status = hmda.model.fi.Parsing
+      updateStatus(status)
+      goto(Parsing) applying SubmissionParsing(Submission(submissionId, status))
+    case Event(_, data) =>
+      stay replying data
+  }
+
+  when(Parsing) {
+    case Event(CompleteParsing, _) =>
+      val status = hmda.model.fi.Parsed
+      updateStatus(status)
+      goto(Parsed) applying SubmissionParsed(Submission(submissionId, status))
+    case Event(CompleteParsingWithErrors, _) =>
+      val status = hmda.model.fi.ParsedWithErrors
+      updateStatus(status)
+      goto(ParsedWithErrors) applying SubmissionParsedWithErrors(Submission(submissionId, status))
+    case Event(_, data) =>
+      stay replying data
+  }
+
+  when(Parsed) {
+    case Event(BeginValidation(_), _) =>
+      val status = hmda.model.fi.Validating
+      updateStatus(status)
+      goto(Validating) applying SubmissionValidating(Submission(submissionId, status))
+    case Event(_, data) =>
+      stay replying data
+  }
+
+  when(ParsedWithErrors) {
+    case Event(_, data) =>
+      stay replying data
+  }
+
+  when(Validating) {
+    case Event(CompleteValidation(_), _) =>
+      val status = hmda.model.fi.Validated
+      updateStatus(status)
+      goto(Validated) applying SubmissionValidated(Submission(submissionId, status))
+    case Event(CompleteValidationWithErrors, _) =>
+      val status = hmda.model.fi.ValidatedWithErrors
+      updateStatus(status)
+      goto(ValidatedWithErrors) applying SubmissionValidatedWithErrors(Submission(submissionId, status))
+    case Event(_, data) =>
+      stay replying data
+  }
+
+  when(Validated) {
+    case Event(GetState, data) =>
+      stay replying data
+  }
+
+  when(ValidatedWithErrors) {
+    case Event(GetState, data) =>
+      stay replying data
+    case Event(_, data) =>
+      stay replying data
+  }
+
+  whenUnhandled {
```

Comment:
I think there might be a misunderstanding here. From the Akka docs: _"we make use of the fact that any event which is not handled by the when() block is passed to the whenUnhandled() block"_. So `whenUnhandled` is sort of a "catch all" for events that might happen in any state, hence why `GetState` is placed here as default. As for the `Event(_, data)` those are present whenever the `SubmissionPersistence` actor is called through the `updateStatus` message, if you follow the code you will see that a response with a `Some(submission)` is returned before the state machine transitions to the next state. Leaving that out causes errors to be logged in the console. Putting it into the `whenUnhandled` block masks the `case Event(e,d)` for catching all other events that might be sent to this state machine by mistake (which I would like to log separately).